### PR TITLE
Bug 2034839: Watch ConfigMap and ImageStream also with the legacy label

### PR DESCRIPTION
**Pull request by akram**
Fixes a regression where watching `ConfigMap` and `ImageStream` was only done on the legacy label.

### CODE REVIEW

**Comment by akram** [comment]
/retest
flake with quay.io or incorrect image version?

**Comment by adambkaplan** [comment]
@akram Looks like we haven't been building this project correctly - if you look at the image build, it is pulling from quay.io instead of referencing an image from the CI build pipeline.

I'm working on a PR in openshift/release to fix this.

**Comment by gabemontero** [comment]
@akram - I *think* your changes may be failing in one of the tests and you'll need to investigate why it is breaking.  From the test output:

```
 === RUN   TestSecretCredentialSyncAfterStartup
    sync_plugin_test.go:108: templateinstance jenkins-sync-plugin-test-namespace-crxf4/jenkins-ephemeral instantiation ready
    sync_plugin_test.go:123: unexpected event type ERROR: &v1.Status{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ListMeta:v1.ListMeta{SelfLink:"", ResourceVersion:"", Continue:"", RemainingItemCount:(*int64)(nil)}, Status:"Failure", Message:"an error on the server (\"unable to decode an event from the watch stream: http2: response body closed\") has prevented the request from succeeding", Reason:"InternalError", Details:(*v1.StatusDetails)(0xc0004aa240), Code:500}
    ref.go:92: full URL http://172.30.209.173:80/credentials/store/system/domain/_/credential/jenkins-sync-plugin-test-namespace-crxf4-secret-to-credential/
    url.go:36: Creating new curl pod curl-cred-pod
    url.go:69: curl pod curl-cred-pod terminated
    sync_plugin_test.go:355: dumpPods looking at pod curl-cred-pod in phase Running
    sync_plugin_test.go:368: pod logs for container hostexec in pod curl-cred-pod:    % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Cancelled
panic: test timed out after 1h15m0s
```

looks like a build ends up cancelled vs. completed ..... more investigation is needed I believe.

If you need help running these tests against one of your personal clusters with these changes, let me know.

**Comment by gabemontero** [comment]
updating the e2e to verify both labels would be ideal as well @akram

File: src/main/java/io/fabric8/jenkins/openshiftsync/ConfigMapClusterInformer.java [hunk 1]
```diff
@@ -53,8 +53,8 @@ public void start() {
         LOGGER.info("Starting cluster wide configMap informer for {} !!" + namespaces);
         LOGGER.debug("listing ConfigMap resources");
         SharedInformerFactory factory = getInformerFactory();
-        Map<String, String> labels = singletonMap(IMAGESTREAM_AGENT_LABEL, IMAGESTREAM_AGENT_LABEL_VALUE);
-        OperationContext withLabels = new OperationContext().withLabels(labels);
+        Map<String, String[]> labels = singletonMap(IMAGESTREAM_AGENT_LABEL, IMAGESTREAM_AGENT_LABEL_VALUES);
```
Comment (gabemontero) [inline src/main/java/io/fabric8/jenkins/openshiftsync/ConfigMapClusterInformer.java:56]: Yeah @akram I suspect this label combination is not achieving the desired result of supporting either label (maybe it is requiring both labels instead) and this is why the existing configmap watch e2e test is failing 

My guess is you'll need to create 2 informers, one for each label individually

Comment (akram) [reply src/main/java/io/fabric8/jenkins/openshiftsync/ConfigMapClusterInformer.java:56]: `kubectl` and also the `informer`  are able to watch on multiple labels values. This change adds this, and the informer should react on any of them (`OR`)  .
I did run the test locally also before the end of year vacations and I was seeing this timeout happening as well and tried to figure out where it was from but with no success.

Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/ConfigMapClusterInformer.java:56]: yeah could be a discrepancy between this java  based version and kubectl

I would suggest looking at the test and mimicking what it does manually so you can debug

Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/ConfigMapClusterInformer.java:56]: > 

I actually misread the test logs @akram 

The test that failed was `TestSecretCredentialSyncAfterStartup` ... the new one Alice added as part of her last fix.

So your change here should have no bearing on secret watches, right?

The only possible common path is the PodTemplateUtils change maybe ... I'll point it out in a sec.

I'm also about to try the specific test locally.

Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/ConfigMapClusterInformer.java:56]: Though I'm still confused as `TestSecretCredentialSyncAfterStartup` does not even start a build (that hangs forever and then cancel

Maybe the tests were running in parallel and that output was from the config map test ??? ... did not think it did that

Yeah I'm going to run the configmap and secret credential tests individually against a cluster with your changes and see.

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java [hunk 2]
```diff
@@ -423,7 +424,8 @@ protected static boolean configMapContainsSlave(ConfigMap configMap) {
 
     protected static boolean hasSlaveLabelOrAnnotation(Map<String, String> map) {
         return map != null && map.containsKey(IMAGESTREAM_AGENT_LABEL)
-                && map.get(IMAGESTREAM_AGENT_LABEL).equals(IMAGESTREAM_AGENT_LABEL_VALUE);
+                && (map.get(IMAGESTREAM_AGENT_LABEL).equals(IMAGESTREAM_AGENT_LABEL_VALUE) ||
+                        map.get(IMAGESTREAM_AGENT_LABEL).equals(IMAGESTREAM_AGENT_ALTERNATE_LABEL_VALUE));
```
Comment (gabemontero) [inline src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java:428]: I forget how order of precedence exactly works here, but you might need to encapsulate within parenthesis the or check on the two possible label values 

i.e. isolate them from the and check with the map null check and presence of the label

File: src/main/java/io/fabric8/jenkins/openshiftsync/ConfigMapClusterInformer.java (hunk 1 above)
Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/ConfigMapClusterInformer.java:56]: Yeah not sure what happened with my copy/paste at https://github.com/openshift/jenkins-sync-plugin/pull/410#issuecomment-1004851896
but with the logs in the latest test run, it is TestConfigMapLegacyPodTemplate that failed

```
=== RUN   TestConfigMapLegacyPodTemplate
    sync_plugin_test.go:108: templateinstance jenkins-sync-plugin-test-namespace-ts2jp/jenkins-ephemeral instantiation ready
    sync_plugin_test.go:210: build phase Pending
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Cancelled
panic: test timed out after 1h15m0s
```

TestSecretCredentialSyncAfterStartup passed, and TestConfigMapLegacyPodTemplate failed for me locally.

The pod template from the config map is not created in the jenkins instance.

I do see these logs in the jenkins logs:

```
2022-01-05 18:28:09 INFO    io.fabric8.jenkins.openshiftsync.PodTemplateUtils purgeTemplates Purging PodTemplates for from Configmap with Uid 26a4f07a-5b95-4e31-ad42-37b834e1bcba
2022-01-05 18:28:09 INFO    io.fabric8.jenkins.openshiftsync.PodTemplateUtils removePodTemplate Removing PodTemplate: jenkins-slave in namespace:  null
2022-01-05 18:28:09 INFO    io.fabric8.jenkins.openshiftsync.PodTemplateUtils removePodTemplate Removing PodTemplate: jenkins-slave in namespace:  null
2022-01-05 18:28:09 INFO    io.fabric8.jenkins.openshiftsync.PodTemplateUtils removePodTemplate Removing PodTemplate: jenkins-slave in namespace:  null
2022-01-05 18:28:09 INFO    io.fabric8.jenkins.openshiftsync.PodTemplateUtils addPodTemplate Adding PodTemplate: jenkins-slave

```

Which makes me think the informer piece work, and that maybe my thought at https://github.com/openshift/jenkins-sync-plugin/pull/410#discussion_r779034595 is the culprit.

I'll stop here and let you pick up from there.

thanks

Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/ConfigMapClusterInformer.java:56]: hmm ... ended up having some more time today, so looked at the code some ... that Adding PodTemplate trace point seems to mean we moved past that label check I was worried about

I'll see if I can add some debug to my copy of your branch.

Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/ConfigMapClusterInformer.java:56]: OK ... shifting gears to your new e2e test case logic.  Your plugin change at least works with the master branch form of `TestConfigMapPodTemplate`

File: test/e2e/sync_plugin_test.go [hunk 3]
```diff
@@ -712,52 +713,73 @@ func TestSecretCredentialSyncAfterStartup(t *testing.T) {
...
 
-	podTemplateTest("sync-plugin-configmap-pod-template", ta)
+func newPodTemplateConfigMap(configMapName string, podTemplateName string, templateLabels map[string]string) *corev1.ConfigMap {
+	templateDefinition := `
+	      <org.csanchez.jenkins.plugins.kubernetes.PodTemplate>
+	        <inheritFrom></inheritFrom>
+	        <name>POD_TEMPLATE_NAME</name>
+	        <instanceCap>2147483647</instanceCap>
+	        <idleMinutes>0</idleMinutes>
+	        <label>jenkins-agent</label>
```
Comment (gabemontero) [inline test/e2e/sync_plugin_test.go:751]:
```suggestion
	        <label>POD_TEMPLATE_NAME</label>
```

File: src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java (hunk 2 above)
Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/PodTemplateUtils.java:428]: yeah ignore this / mark resolved

File: src/main/java/io/fabric8/jenkins/openshiftsync/ConfigMapClusterInformer.java (hunk 1 above)
Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/ConfigMapClusterInformer.java:56]: Yeah this multi label stuff is OK too

With my suggestion at https://github.com/openshift/jenkins-sync-plugin/pull/410#pullrequestreview-845082555 and running with your branch for this things work.

File: test/e2e/sync_plugin_test.go (hunk 3 above)
Comment (akram) [reply test/e2e/sync_plugin_test.go]: nice catch @gabemontero ; I was thinking that only the name was important here and I let this label as is for that. But, when the test broke  I forgot to return to it and fix it.

File: src/main/java/io/fabric8/jenkins/openshiftsync/ConfigMapClusterInformer.java (hunk 1 above)
Comment (akram) [reply src/main/java/io/fabric8/jenkins/openshiftsync/ConfigMapClusterInformer.java:56]: thanks @gabemontero , I just commited that fix.

**Comment by gabemontero** [comment]
The config map pod template e2e's are not OK @akram  but `TestImageStreamPodTemplate` did not complete but with a similar error profile as the config map pod templates

Hopefully a similar error in the new e2e's there ... taking a look now

**Comment by akram** [comment]
> TestImageStreamPodTemplate

OK, got it now. The refactor I have done takes the podTemplateName as the first argument in `podTemplateTest()`
And, for this test, I am passing an incorrect value. I will fix that and find others.
`podTemplateTest("sync-plugin-imagestream-pod-template", ta)`

**Comment by gabemontero** [comment]
OK this time all the ConfigMap and ImageStream pod template tests passed .... the last one never completed before the test  time out .... I wonder if we just hit the current 75 minute limit this time and we need to bump that up.

I'll check out the latest and try the failed test on its own locally, and make sure it was not inadvertently broke by the other changes

For reference: 

```
=== RUN   TestPruningFailedPipeline
    sync_plugin_test.go:108: templateinstance jenkins-sync-plugin-test-namespace-xpngl/jenkins-ephemeral instantiation ready
    sync_plugin_test.go:123: unexpected event type ERROR: &v1.Status{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ListMeta:v1.ListMeta{SelfLink:"", ResourceVersion:"", Continue:"", RemainingItemCount:(*int64)(nil)}, Status:"Failure", Message:"an error on the server (\"unable to decode an event from the watch stream: http2: response body closed\") has prevented the request from succeeding", Reason:"InternalError", Details:(*v1.StatusDetails)(0xc00010e7e0), Code:500}
    sync_plugin_test.go:210: build phase Pending
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Pending
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Pending
    sync_plugin_test.go:210: build phase Pending
    sync_plugin_test.go:210: build phase Running
    sync_plugin_test.go:210: build phase Pending
    sync_plugin_test.go:210: build phase Pending
    sync_plugin_test.go:210: build phase Pending
```

**Comment by gabemontero** [comment]
> OK this time all the ConfigMap and ImageStream pod template tests passed .... the last one never completed before the test time out .... I wonder if we just hit the current 75 minute limit this time and we need to bump that up.
> 
> I'll check out the latest and try the failed test on its own locally, and make sure it was not inadvertently broke by the other changes
> 
> For reference:
> 
> ```
> === RUN   TestPruningFailedPipeline
>     sync_plugin_test.go:108: templateinstance jenkins-sync-plugin-test-namespace-xpngl/jenkins-ephemeral instantiation ready
>     sync_plugin_test.go:123: unexpected event type ERROR: &v1.Status{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ListMeta:v1.ListMeta{SelfLink:"", ResourceVersion:"", Continue:"", RemainingItemCount:(*int64)(nil)}, Status:"Failure", Message:"an error on the server (\"unable to decode an event from the watch stream: http2: response body closed\") has prevented the request from succeeding", Reason:"InternalError", Details:(*v1.StatusDetails)(0xc00010e7e0), Code:500}
>     sync_plugin_test.go:210: build phase Pending
>     sync_plugin_test.go:210: build phase Running
>     sync_plugin_test.go:210: build phase Pending
>     sync_plugin_test.go:210: build phase Running
>     sync_plugin_test.go:210: build phase Pending
>     sync_plugin_test.go:210: build phase Pending
>     sync_plugin_test.go:210: build phase Running
>     sync_plugin_test.go:210: build phase Pending
>     sync_plugin_test.go:210: build phase Pending
>     sync_plugin_test.go:210: build phase Pending
> ```

Interesting ... I was able to reproduced this locally, though it is not obvious to me how your changes cause it. 

The last of the valid build (number 4) is still in pending state if you look at it from openshift:

```
gmontero ~ $ oc projects | grep jenkins | xargs -l -r oc get builds -n 
NAME                                TYPE              FROM   STATUS    STARTED         DURATION
sync-plugin-prune-failed-builds-2   JenkinsPipeline          Failed    8 minutes ago   
sync-plugin-prune-failed-builds-3   JenkinsPipeline          Failed    8 minutes ago   
sync-plugin-prune-failed-builds-4   JenkinsPipeline          Pending                   

```

But if you log onto the jenkins console, that build completed as failed, as expected

So something happened in sync plugin where it failed to move build 4 from pending to running, then failed.

I don't see any obvious errors in the jenkins pod logs.

I then reverted the plugin changes from this PRs branch so that the jenkins image ran with the plugin version that is currently in the 4.10 image, but ran the tests from your branch.  The test passed successfully multiple times.  I even paused the test before it exited in the debugger so I could inspect everything, and everything looked good.  The right number of builds existed in the right state.

It certainly is not obvious to me *why* your pod template related changes broke pruning, but that seems to be the case.  

See if you can debug this Friday morning your time, and then we can sync up as needed after the demos on Friday, and go from there.

**Comment by gabemontero** [comment]
Yeah it appears to be intermittent ... just switched back to your updates to both the plugin and tests and now it passed.

**Comment by akram** [comment]
@gabemontero I will continue to have a look at it tomorrow. I did also the tests on my local cluster and didn't see the regression to happen. More over, the pruning test doesn't even use a `PodTemplate` as its pipeline definition does not specify any `node()` so it runs on the master.

So, the way you described it, is more likely to be a synchronisation issue while update the build to job map. I am re-runing the tests; they may pass if it was like condition. If it is the case, I will have investigate this point deeper. But, as this part is untouched between 1.0.51 and now, it also probably means that 1.0.51 has the issue.

**Comment by gabemontero** [comment]
> @gabemontero I will continue to have a look at it tomorrow. I did also the tests on my local cluster and didn't see the regression to happen. More over, the pruning test doesn't even use a `PodTemplate` as its pipeline definition does not specify any `node()` so it runs on the master.
> 
> So, the way you described it, is more likely to be a synchronisation issue while update the build to job map. I am re-runing the tests; they may pass if it was like condition. If it is the case, I will have investigate this point deeper. But, as this part is untouched between 1.0.51 and now, it also probably means that 1.0.51 has the issue.

Yeah I'll run a few more times and at least get some sense of the frequency ... it may be you changes somehow have tweaked the timing window to make it more readily recurrent.

**Comment by gabemontero** [comment]
AWS flake

/retest

**Comment by gabemontero** [comment]
Yeah I've run it a dozen times locally now with no errors.

Probably a question @akram if the timing window is widest on the initial running.  We'll see what happens with your e2e here if we get past aws flakes.

**Pull request merged**